walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,110 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
-
+import java.util.Collections;
+/* Implementation requires that Collection of Events are all sorted in order of meeting start time */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> availableTimes = new ArrayList<>();
+
+    /* If the requested time span is longer than a whole day, return no available meeting times */
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return availableTimes;
+    }
+
+    int blockStart = TimeRange.START_OF_DAY;
+    int blockStartOpt = TimeRange.START_OF_DAY;
+    Collection<String> requestAtten = request.getAttendees();
+    Collection<String> optionalAtten = request.getOptionalAttendees();
+
+    /* If there are only optional attendees requested, default set all optional attendeeds to be included */
+    boolean optionalIncluded = request.getAttendees().isEmpty();
+    
+    for (Event event : events) { 
+      /* Check if there are attendees in the event that are also requested attendees */
+      Collection<String> eventAtten = event.getAttendees();
+      boolean timeConflictFound = !Collections.disjoint(eventAtten, requestAtten);
+      boolean timeConflictFoundOpt = !Collections.disjoint(eventAtten, optionalAtten);
+      int eventStart = event.getWhen().start();
+      int eventEnd = event.getWhen().end();
+
+      /* If no optional attendee conflict and event length long enough, include optional attendee */
+      if (!timeConflictFoundOpt && eventStart - blockStartOpt >= request.getDuration()) {
+        optionalIncluded = true;
+      }
+      else if (timeConflictFoundOpt) {
+        blockStartOpt = eventEnd;
+      }
+
+      if (optionalIncluded) {
+        timeConflictFound =  timeConflictFound || timeConflictFoundOpt;
+      }
+
+      if (timeConflictFound) {
```

Comment:
It felt really sus to me that you only add to availableTimes IF a conflict is found, which feels inverted. After about an hour I finally came up with a test case to break this:

```java
  @Test
  public void exploit() {
    // Have a required attendee A with no meetings. Use an optional attendee to
    // exploit a bug.
    //
    // Events  : |----O----| |--O---|
    // Day     : |------------------|
    // Options :           |-|
    
    Collection<Event> events = Arrays.asList(
        new Event("OptionalIgnored",
                  TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0900AM, false),
            Arrays.asList(PERSON_O)),
        new Event("OptionalIgnored2",
                  TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, false),
            Arrays.asList(PERSON_O)));
    
    MeetingRequest request = new MeetingRequest(Arrays.asList(PERSON_A), DURATION_30_MINUTES);
    request.addOptionalAttendee(PERSON_O);

    Collection<TimeRange> actual = query.query(events, request);
    Collection<TimeRange> expected =
        Arrays.asList(TimeRange.fromStartEnd(TIME_0900AM, TIME_0930AM, false));

    Assert.assertEquals(expected, actual);
  }
```

Now I think there's a larger issue - I spent around 40 minutes of that hour trying to understand how this algorithm works and I still don't understand how it works. I think the code either needs to be structured to work and/or there needs to be a comment explaining how the algorithm works. There's way too many variables (availableTimes, blockStart, blockEnd, blockStartOpt, timeConflictFound, timeConflictFoundOpt, eventStart, eventEnd, optionalIncluded...) for a human to keep track of at any given time.

Reply:
Given that this test demonstrates a bug that you've detected and fixed, I think it'd be worthwhile to add this to your test cases.

Reply:
I did, I just renamed it to freeAllDay rather than exploit in the test file.

Reply:
I'm dumb ignore me :)